### PR TITLE
Gevent

### PR DESCRIPTION
**Pull request by Elfoniok**

### CODE REVIEW

File: golem/node.py [hunk 1]
```diff
@@ -60,6 +60,9 @@ def run(self, use_rpc=False):
             if is_windows() and 'twisted.internet.reactor' not in sys.modules:
                 from twisted.internet import iocpreactor
                 iocpreactor.install()
+            elif 'twisted.internet.reactor' not in sys.modules:
+            	from golem.reactor import geventreactor
+            	geventreactor.install()
```
Comment (jiivan) [inline golem/node.py]:
```
golem/node.py:64:13: E101 indentation contains mixed spaces and tabs
golem/node.py:64:13: W191 indentation contains tabs
golem/node.py:65:13: E101 indentation contains mixed spaces and tabs
golem/node.py:65:13: W191 indentation contains tabs
```

File: gui/startapp.py [hunk 2]
```diff
@@ -2,6 +2,9 @@
 import os
 import subprocess
 import sys
+from os import path
+from twisted.internet.defer import setDebugging
+from twisted.internet.error import ReactorAlreadyRunning
```
Comment (jiivan) [inline gui/startapp.py]:
```
gui/startapp.py:5:1: F401 'os.path' imported but unused
gui/startapp.py:6:1: F401 'twisted.internet.defer.setDebugging' imported but unused
gui/startapp.py:16:1: F811 redefinition of unused 'ReactorAlreadyRunning' from line 7
```

File: gui/startapp.py [hunk 3]
```diff
@@ -59,7 +62,13 @@ def start_client(start_ranking, datadir=None,
         if is_windows():
             from twisted.internet import iocpreactor
             iocpreactor.install()
+        else:
+	    from golem.twisted.reactor import geventreactor
+            geventreactor.install()
```
Comment (jiivan) [inline gui/startapp.py]:
```
gui/startapp.py:66:1: E101 indentation contains mixed spaces and tabs
gui/startapp.py:66:1: W191 indentation contains tabs
gui/startapp.py:67:1: E101 indentation contains mixed spaces and tabs
```

File: gui/startapp.py [hunk 4]
```diff
@@ -59,7 +62,13 @@ def start_client(start_ranking, datadir=None,
         if is_windows():
             from twisted.internet import iocpreactor
             iocpreactor.install()
+        else:
+	    from golem.twisted.reactor import geventreactor
+            geventreactor.install()
         from twisted.internet import reactor
+    from golem.rpc.router import CrossbarRouter
```
Comment (jiivan) [inline gui/startapp.py]:
```
gui/startapp.py:75:5: F811 redefinition of unused 'CrossbarRouter' from line 69
```

File: tests/gui/test_startapp.py [hunk 5]
```diff
@@ -67,6 +67,7 @@ def _start_client(self, router_fails=False, session_fails=False, expected_result
         queue = Queue()
 
         with patch('gui.startapp.start_gui'), \
+             patch('golem.reactor.geventreactor.install'), \
```
Comment (jiivan) [inline tests/gui/test_startapp.py]: Probably better use decorator syntax like this:
```python
@patch('gui.startapp.start_gui')
@patch('golem.reactor.geventreactor.install')
# [...] More all other decorators
def _start_client(self, router_fails=False, session_fails=False, expected_result=None, *mocks):
```


```
tests/gui/test_startapp.py:70:14: E127 continuation line over-indented for visual indent
tests/gui/test_startapp.py:71:14: E127 continuation line over-indented for visual indent
tests/gui/test_startapp.py:72:14: E127 continuation line over-indented for visual indent
tests/gui/test_startapp.py:73:14: E127 continuation line over-indented for visual indent
tests/gui/test_startapp.py:74:14: E127 continuation line over-indented for visual indent
tests/gui/test_startapp.py:75:14: E127 continuation line over-indented for visual indent
```

**Review by jiivan — CHANGES_REQUESTED** [review]
`flake8 --ignore=e501` errors

File: golem/node.py (hunk 1 above)
Comment (Elfoniok) [reply golem/node.py]: Ok

File: gui/startapp.py (hunk 2 above)
Comment (Elfoniok) [reply gui/startapp.py]: Ok

File: gui/startapp.py (hunk 4 above)
Comment (Elfoniok) [reply gui/startapp.py]: Ok

File: tests/gui/test_startapp.py (hunk 5 above)
Comment (Elfoniok) [reply tests/gui/test_startapp.py]: Please explain it in more details.

File: gui/startapp.py (hunk 3 above)
Comment (Elfoniok) [reply gui/startapp.py]: Ok

File: tests/gui/test_startapp.py (hunk 5 above)
Comment (jiivan) [reply tests/gui/test_startapp.py]:
```diff
diff --git a/tests/gui/test_startapp.py b/tests/gui/test_startapp.py
index 740598a..2a287cd 100644
--- a/tests/gui/test_startapp.py
+++ b/tests/gui/test_startapp.py
@@ -62,18 +62,17 @@ class TestStartAppFunc(TestDirFixtureWithReactor):
         assert len(envs) >= 2
 
     def _start_client(self, router_fails=False, session_fails=False, expected_result=None):
-
-        client = None
         queue = Queue()
 
-        with patch('gui.startapp.start_gui'), \
-             patch('golem.reactor.geventreactor.install'), \
-             patch('golem.client.Client.start', side_effect=lambda *_: queue.put(u"Success")), \
-             patch('golem.client.Client.sync'), \
-             patch('gui.startapp.start_error', side_effect=lambda err: queue.put(err)), \
-             patch('golem.rpc.router.CrossbarRouter.start', router_start(router_fails)), \
-             patch('golem.rpc.session.Session.connect', session_connect(session_fails)):
-
+        @patch('gui.startapp.start_gui')
+        @patch('golem.reactor.geventreactor.install')
+        @patch('golem.client.Client.start', side_effect=lambda *_: queue.put(u"Success"))
+        @patch('golem.client.Client.sync')
+        @patch('gui.startapp.start_error', side_effect=lambda err: queue.put(err))
+        @patch('golem.rpc.router.CrossbarRouter.start', router_start(router_fails))
+        @patch('golem.rpc.session.Session.connect', session_connect(session_fails))
+        def inner(*mocks):
+            client = None
             try:
                 client = Client(datadir=self.path,
                                 transaction_system=False,
@@ -89,12 +88,12 @@ class TestStartAppFunc(TestDirFixtureWithReactor):
 
                 message = queue.get(True, 10)
                 assert unicode(message).find(expected_result) != -1
-
             except Exception as exc:
                 self.fail(u"Cannot start client process: {}".format(exc))
             finally:
                 if client:
                     client.quit()
+        return inner()
 
     def _start_gui(self, session_fails=False, expected_result=None):
 ```

**Review by jiivan — APPROVED** [review]
LGTM

**Comment by paulperegud** [comment]
On fresh virtualenv environment crashes with following when I launch `golemapp`:
```
Traceback (most recent call last):
  File "/home/pawel/.virtualenvs/golem-gevent/local/lib/python2.7/site-packages/crossbar/controller/node.py", line 478, in start
    yield self._startup(self._config)
AssertionError
Unhandled Error
Traceback (most recent call last):
  File "/home/pawel/code/ggolem/golem/golemapp.py", line 51, in start
    start_app(rendering=True, **config)
  File "/home/pawel/code/ggolem/golem/gui/startapp.py", line 120, in start_app
    transaction_system, **config_overrides)
  File "/home/pawel/code/ggolem/golem/gui/startapp.py", line 108, in start_client
    reactor.run()
  File "/home/pawel/.virtualenvs/golem-gevent/local/lib/python2.7/site-packages/twisted/internet/base.py", line 1243, in run
    self.mainLoop()
--- <exception caught here> ---
  File "/home/pawel/code/ggolem/golem/golem/reactor/geventreactor.py", line 369, in mainLoop
    c()
  File "/home/pawel/code/ggolem/golem/golem/reactor/geventreactor.py", line 192, in __call__
    self.func(*self.a, **self.kw)
  File "/home/pawel/.virtualenvs/golem-gevent/local/lib/python2.7/site-packages/gevent/hub.py", line 210, in kill
    if not greenlet.dead:
exceptions.AttributeError: 'NoneType' object has no attribute 'dead'
```

**Comment by paulperegud** [comment]
LGTM

**Review by paulperegud — APPROVED** [review]

File: tests/golem/test_opt_node.py [hunk 6]
```diff
@@ -100,7 +103,8 @@ def test_node_address_missing(self):
 
     @ci_skip
     @patch('golemapp.OptNode')
-    def test_single_peer(self, mock_node):
+    @patch('golem.core.common.config_logging')
```
Comment (badb) [inline tests/golem/test_opt_node.py:106]: Can you add this patch also to test_many_peers?

**Review by badb — APPROVED** [review]

File: tests/golem/test_opt_node.py (hunk 6 above)
Comment (Elfoniok) [reply tests/golem/test_opt_node.py:106]: Done

**Pull request closed without merging**